mitiq/zne.py

```diff
@@ -1,50 +1,122 @@
-from typing import List, Tuple, Callable
+"""Zero-noise extrapolation tools."""
+
+from typing import Callable
 import mitiq.qiskit.qiskit_utils as qs_utils
 from mitiq import QPROGRAM
 from mitiq.factories import Factory, LinearFactory
 
 
-def mitigate(pq: QPROGRAM, fac: Factory, scale_noise: Callable[[QPROGRAM, float], QPROGRAM], run_program: Callable) -> \
-        Tuple[List[float], List[float]]:
+def run_factory(fac: Factory, noise_to_expval: Callable[[float], float]) -> None:
     """
-    Runs the factory until convergence and returns the full list of noise values and expectations that have been
-    calculated.
-    :param pq: Program to mitigate.
+    Runs a factory until convergence and returns the final zero-noise extrapolation.
+
+    Args:
+        fac: Instance of Factory object to be run.
+        noise_to_expval: Function mapping noise scale values to expectation vales.
     """
-    if not fac.is_converged(fac.instack, fac.outstack):
-        next_param = fac.step(fac.instack, fac.outstack)
+    if not fac.is_converged():
+        next_param = fac.next()
+        next_result = noise_to_expval(next_param)
+        fac.push(next_param, next_result)
+
+        return run_factory(fac, noise_to_expval)
```

Comment:
Why not just use an iterative approach here instead of recursive? In my mind, iterative is simpler to see what's happening here, but also practically the call stack could get quite large and slow down performance.

For example, an iterative approach would be something like:

```python
while not fac.is_converged() and iterations < max_iterations:
    next_param = fac.next()
    next_results = noise_to_expval(next_param)
    fac.push(nex_param, next_result)
    iterations += 1
```

Reply:
Done as suggested. I also added `max_iterations` as an optional argument and a warning message in the case in which `max_iterations` is reached.